mongock-core/mongock-runner/mongock-runner-core/src/main/java/io/mongock/runner/core/executor/ChangeLogRuntimeImpl.java

```diff
@@ -120,6 +124,29 @@ private String getParameterName(Parameter parameter) {
 
 
   private Constructor<?> getConstructor(Class<?> type) {
-    return type.getConstructors()[0];
+    return findChangeUnitConstructor(type)
+        .orElseGet(() -> findDefaultConstructor(type));
+  }
+
+  private Constructor<?> findDefaultConstructor(Class<?> type) {
+    Supplier<Stream<Constructor<?>>> constructorSupplier = () -> Arrays.stream(type.getConstructors());
+    if (constructorSupplier.get().count() > 1) {
+      logger.warn("Mongock found multiple constructors for changeUnit[{}]. " +
+          "It's recommended to annotate the one you want Mongock to use with @ChangeUnitConstructor. " +
+          "FROM VERSION 6 THIS WILL CAUSE AN ERROR ", type.getName());
+    }
+    return constructorSupplier.get()
+        .findFirst()
+        .orElseThrow(() -> new MongockException("Mongock cannot find a valid constructor for changeUnit[%s]", type.getName()));
+  }
+
+  private Optional<Constructor<?>> findChangeUnitConstructor(Class<?> type) {
+    Supplier<Stream<Constructor<?>>> changeUnitConstructorsSupplier = () -> Arrays.stream(type.getConstructors())
+        .filter(constructor -> constructor.isAnnotationPresent(ChangeUnitConstructor.class));
+    if (changeUnitConstructorsSupplier.get().count() > 1) {
+      throw new MongockException("Found multiple constructors for changeUnit[%s] without annotation @ChangeUnitConstructor." +
+          " Annotate the one you want Mongock to use to instantiate your changeUnit", type.getName());
+    }
+    return changeUnitConstructorsSupplier.get().findFirst();
   }
 }
```

Comment:
It's also fine, but there is no need for Supplier. 
What about:
```java
  private Optional<Constructor<?>> findChangeUnitConstructor(Class<?> type) {
    Stream<Constructor<?>> constructorStream = Arrays.stream(type.getConstructors())
        .filter(constructor -> constructor.isAnnotationPresent(ChangeUnitConstructor.class));
    if (constructorStream.count() > 1) {
      throw new MongockException("Found multiple constructors for changeUnit[%s] without annotation @ChangeUnitConstructor." +
          " Annotate the one you want Mongock to use to instantiate your changeUnit", type.getName());
    }
    return constructorStream.findFirst();
  }
```

Reply:
This won't work as the `(constructorStream.count()`  ends stream. I have tests that'd fail with the recommended changes.